### ISSUE TITLE
Performance Issues

### ISSUE THREAD

**Issue by NecOnIce** [habanero-rice/habanero-upc#2]
Hi,

i have compiled HUPCPP with Gasnet and ibv-conduit on an Linux Cluster with Infiniband Network. With this Configuration i have run the following program:

```
//
// Created by jonas on 28.02.16.
//

#define GASNET_SEQ

#include <fftw3.h>
#include "hcupc_spmd.h"
#include <chrono>
#include <random>

using namespace std;

typedef struct {
    double real;
    double imag;
} complex_t;

upcxx::global_ptr<complex_t> *input, *output;
upcxx::global_ptr<upcxx::global_ptr<complex_t> > allWeights;
upcxx::global_ptr<complex_t> *weights;
size_t rows, columns;
#define ROWS_DEFAULT 1024

void fillRandom(complex_t *arr, size_t rows, size_t columns) {

    // create random generator
    default_random_engine generator;
    uniform_real_distribution<double > distribution(0.0f, 1.0f);

    for (size_t row = 0; row < rows; row++) {
        for (size_t column = 0; column < columns; column++) {
            size_t index = row*columns+column;
            arr[index].real = distribution(generator);
            arr[index].imag = distribution(generator);
        }
    }
}

void runFFT(size_t rowsPerPlace, complex_t *myInput, complex_t *myOutput) {

    int num[] = {columns};
    fftw_plan plan = fftw_plan_many_dft(1, num, rowsPerPlace, (fftw_complex *)myInput, NULL, 1, columns,
                                        (fftw_complex *)myOutput,NULL, 1, columns, FFTW_FORWARD, FFTW_ESTIMATE);

    fftw_execute(plan);
    fftw_destroy_plan(plan);
    fftw_cleanup();
}

void runFFT2(complex_t *input, complex_t *output) {

    fftw_plan plan = fftw_plan_dft_2d(rows, columns, (fftw_complex *)input, (fftw_complex *)output,
                                      FFTW_FORWARD, FFTW_ESTIMATE);
    fftw_execute(plan);
    fftw_destroy_plan(plan);
    fftw_cleanup();
}

void transpose(complex_t *in, complex_t *out, size_t rows, size_t columns) {

    for (size_t i = 0; i < rows; i++) {
        for (size_t j = 0; j < columns; j++) {
            out[j*rows+i].real = in[i*columns+j].real;
            out[j*rows+i].imag = in[i*columns+j].imag;
        }
    }
}

void fft(size_t rows, size_t columns) {

    int places = upcxx::ranks();
    size_t myID = upcxx::myrank();

//    cout << "fft run on rank: " << myID << endl;

    size_t rowsPerPlace = rows / places;
    size_t columnsPerPlace = columns / places;

    complex_t *myInput = input[myID].raw_ptr();
    complex_t *myOutput = output[myID].raw_ptr();

    assert(rows == columns); // only handle square case for now

    // generate random data for the 2-D input array
    fillRandom(input[myID].raw_ptr(), rowsPerPlace, columns);

    // run fft in first dimension
    runFFT(rowsPerPlace, myInput, myOutput);

//    cout << "runFFT finished." << endl;

    // copy allWeights from place 0 to local asynchronous
    upcxx::global_ptr<complex_t> *localAllWeights = new upcxx::global_ptr<complex_t> [places];
    upcxx::async_copy(allWeights, upcxx::global_ptr<upcxx::global_ptr<complex_t> > (localAllWeights), places);

//    cout << "localWeights copied." << endl;

    // transpose Output into temporary Variable
    complex_t *temp = new complex_t[rowsPerPlace*columns];
    transpose(myOutput, temp, rowsPerPlace, columns);

//    cout << "first transpose finished" << endl;

    // transpose back the values from temp into Output for each other Place
    size_t msgsPerPlace = rowsPerPlace * columnsPerPlace;
    for (size_t i = 0; i < places; i++) {
        transpose(temp + i*msgsPerPlace, myOutput + i*msgsPerPlace, columnsPerPlace, rowsPerPlace);
    }

//    cout << "second transpose finished" << endl;

    // synchronize all asynchronous copies
    upcxx::async_copy_fence();

    // exchange the weights
    for (size_t i = 0; i < places; i++) {
        upcxx::async_copy(myOutput+i*msgsPerPlace, localAllWeights[i] + myID*msgsPerPlace, msgsPerPlace);
    }

    // synchronize all asynchronous copies
    upcxx::async_copy_fence();

//    cout << "weights exchanged" << endl;

    // transpose weights
    transpose(localAllWeights[myID].raw_ptr(), temp, columns, rowsPerPlace);

//    cout << "thrid transpose finished" << endl;

    // run fft in second dimension
    runFFT(rowsPerPlace, temp, myOutput);

//    cout << "last fft finished" << endl;

    // free memory of temp
    delete[](temp);
}

int main(int argc, char **argv)
{

    int tmp = 0;
    // parsing the command line arguments
    if (argc > 2) {
        tmp = atoi(argv[2]);
    }

    rows = (tmp < 2) ? ROWS_DEFAULT : tmp;
    columns = rows;

    hupcpp::launch(&argc, &argv, [] () {

        size_t places = upcxx::ranks();
        input = new upcxx::global_ptr<complex_t> [places];
        output = new upcxx::global_ptr<complex_t> [places];
        allWeights = upcxx::allocate<upcxx::global_ptr<complex_t> > (upcxx::myrank(), places);
        weights = allWeights.raw_ptr();
        size_t rowsPerPlace = rows / places;

        cout << "run fft2d with: " << places << " Places." << endl;

        // allocate memory with 1-D data partitioning
        for (size_t i = 0; i < places; i++) {
            input[i] = upcxx::allocate<complex_t>(i, rowsPerPlace*columns);
            output[i] = upcxx::allocate<complex_t>(i, rowsPerPlace*columns);
            weights[i] = upcxx::allocate<complex_t>(i, rowsPerPlace*columns);
        }

        auto startTime = chrono::high_resolution_clock::now();

        hupcpp::finish_spmd([places] () {

            // perform fft in parallel
            for (int i = 0; i < places; i++) {

                hupcpp::asyncAt(i, [] () {
                    fft(rows, columns);
                });
            }
        });

        auto endTime = chrono::high_resolution_clock::now();

        cout << "Wall clock time passed: " << chrono::duration<double, milli>(endTime - startTime).count()
        << " ms" << endl;

        // releasing place memory
        for (int i = 0; i < places; i++) {
            upcxx::deallocate(input[i]);
            upcxx::deallocate(output[i]);
            upcxx::deallocate(weights[i]);
        }

//        cout << "Memory freed" << endl;
        delete [] input;
        delete [] output;
    });



    return 0;
}
```

Bit if i use more than 1 Node for execution, the performance becomes very bad. On the Cluster i have  49406.4 ms for rows = columns = 4096. On my own Computer with the MPI-Conduit i have a time of 16411 ms. If i use the MPI Conduit on the Linux Cluster, the Performance is more worse...

I have implemented the same Program for the APGAS Framework in Java and with APGAS the time is 9629 ms for one Place and rows = columns = 4096 and decreases for each more Place i supply to 5380 ms with 16 Places.

Is there another issue at HUPCPP or inside the HCLIB ? Or have i used a wrong configuration of Gasnet and UPCXX ?

**Comment by agrippa** [comment]
@NecOnIce Performance problems are always hard to debug remotely. One of the items I'm actively working on is a performance regression suite for HClib to help diagnose exactly these types of issues, maybe HUPC will be the next project I look at. Do you mind also sharing your APGAS implementation? It might be interesting to see if we can reproduce your issue on our own clusters.

**Comment by NecOnIce** [comment]
@agrippa Sry for my very late response. I have attached the sources i have produced for my testings. I hope you get all running. The APGAS implementations are build with gradle. I have also provided my version of the APGAS library to the zip archive, to avoid failures which can happen if another version of APGAS is used.

My Timings for the programs i have written are summed up in this file:
[Timings.pdf](https://github.com/habanero-rice/habanero-upc/files/283876/Timings.pdf)

Additionally i think the performance issues are caused by the work-stealing algorithm. I think to much steals slows down the performance, because the cpu's are used for steals most of the time. 

[APGAS.zip](https://github.com/habanero-rice/habanero-upc/files/283847/APGAS.zip)
[HabaneroUPC++.zip](https://github.com/habanero-rice/habanero-upc/files/283848/HabaneroUPC.zip)
